pkg/controllers/updaterun/execution.go

```diff
@@ -85,6 +94,87 @@ func (r *Reconciler) execute(
 	return finished, clusterUpdatingWaitTime, execErr
 }
 
+// checkBeforeStageTasksStatus checks if the before stage tasks have finished.
+// It returns if the before stage tasks have finished or error if the before stage tasks failed.
+// It also returns the time to wait before rechecking the wait type of task. It turns -1 if the task is not a wait type.
+func (r *Reconciler) checkBeforeStageTasksStatus(ctx context.Context, updatingStageIndex int, updateRun placementv1beta1.UpdateRunObj) (bool, error) {
+	updateRunRef := klog.KObj(updateRun)
+	updateRunStatus := updateRun.GetUpdateRunStatus()
+	updatingStageStatus := &updateRunStatus.StagesStatus[updatingStageIndex]
+	updatingStage := &updateRunStatus.UpdateStrategySnapshot.Stages[updatingStageIndex]
+	if updatingStage.BeforeStageTasks == nil {
+		klog.V(2).InfoS("There is no before stage task for this stage", "stage", updatingStage.Name, "updateRun", updateRunRef)
+		return true, nil
+	}
+	passed := true
+	for i, task := range updatingStage.BeforeStageTasks {
+		switch task.Type {
+		case placementv1beta1.StageTaskTypeApproval:
+			beforeStageTaskApproved := condition.IsConditionStatusTrue(meta.FindStatusCondition(updatingStageStatus.BeforeStageTaskStatus[i].Conditions, string(placementv1beta1.StageTaskConditionApprovalRequestApproved)), updateRun.GetGeneration())
```

Comment:
Looks like most of code is the same for both checkAfterStageTasksStatus & checkBeforeStageTasksStatus if task is of type approval, lets move the code to a common function so we can reuse